### PR TITLE
WIP: do not merge, Feature expose precision loss logic

### PR DESCRIPTION
**Pull request by hmaarrfk**
## Description
I'm working to convert images between raw with bayer pattern and rgb or grey scale. It would be good to have this logic exposed to avoid unnecessary image copies.

This logic will be reused there to catch most cases.

In writing the tests for this, I found Issue #3080. This PR fixes Issue #3080.

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
I also just added optional warnings.

I guess warnings are nice the first time you run your code, but after a while you realize that you are OK with the precision loss and therefore they shouldn't add to visual noise if you don't need them.

**Comment by stefanv** [comment]
Decision: refactor dtype range checking; will not include warning handling.

**Comment by hmaarrfk** [comment]
@scikit-image/core bump

File: skimage/util/dtype.py [hunk 1]
```diff
@@ -32,7 +32,73 @@
...
+    dtypeobj_in = np.dtype(dtypeobj_in)
+    dtypeobj_out = np.dtype(dtypeobj_out)
+    kind_in = dtypeobj_in.kind
+    kind_out = dtypeobj_out.kind
+    itemsize_in = dtypeobj_in.itemsize
+    itemsize_out = dtypeobj_out.itemsize
+
+    has_loss = False
+
+    if kind_in != 'b' and kind_out == 'b':
```
Comment (stefanv) [inline skimage/util/dtype.py]: There seems to be some more generic pattern here.

Can we check only the itemsize_in vs itemsize_out?

Comment (stefanv) [reply skimage/util/dtype.py]: plus minor logic for signed ranges

Comment (hmaarrfk) [reply skimage/util/dtype.py]: seems like i can just treat it as 'u' and do something like `kind in ['b', 'u']`

Comment (hmaarrfk) [reply skimage/util/dtype.py]: Seems like I cannot.
```
In [1]: import numpy as np

In [2]: np.dtype(np.uint8).itemsize
Out[2]: 1
```

Comment (stefanv) [reply skimage/util/dtype.py]: So `bool` is the exception that needs to be checked for, because it is atypically sized.

**Comment by hmaarrfk** [comment]
@stefanv commenting here so this discussion doesn't get lost

right, so that exception checking is done in the line you pointed to right?

the last 3 if statements do the boundary checking for the different int conversions. A uint8->int8 actually has loss, so that is why I use the appropriate comparison operator.

I think the float logic was written assuming 32 bit floats. In which case, they are more expressive than the other types we support. Can I access things like the number of bits in the mantisa? that is the correct thing to compare.


The logic for the

**Comment by hmaarrfk** [comment]
I think I have an idea. Let me think about it later tonight.

**Comment by stefanv** [comment]
Yes!  That looks great.

**Comment by hmaarrfk** [comment]
merger: please squash merge

**Comment by hmaarrfk** [comment]
@stefanv I think I was off by one in the last version you saw:

the current version uses:
```
 59     def n_significant_bits(t):                                                  
 60         dtypeobj = np.dtype(t)                                                  
 61         kind = dtypeobj.kind                                                    
 62         if kind == 'f':                                                         
 63             # IEEE floating point point has an implied 1, increasing the number 
 64             # of significant bits given a mantisa of a certain size                                                                                                      
 65             return np.finfo(dtypeobj).nmant + 1
```
which i think is more correct. though, if you are relying on 1 very small bit, maybe your algorithm is too sensitive.

File: skimage/util/dtype.py [hunk 2]
```diff
@@ -32,7 +32,68 @@
...
+        float -> signed/unsigned ints will always return true.
+        signed/unsigned -> float of same number of bytes will return true.
+
+    Parameters
+    ----------
+    dtypeobj_in: np.dtype
+        dtype of the input image.
+    dtypeobj_out: np.dtype
+        dtype of the output image.
+    output_warning: bool, optional
```
Comment (sciunto) [inline skimage/util/dtype.py]: This and the option below do not appear in the signature

File: skimage/util/dtype.py [hunk 3]
```diff
@@ -32,7 +32,68 @@
...
+        has_loss = False
+
+    if has_loss:
+        warn("Possible precision loss when converting from {} to {}"
+             .format(dtypeobj_in, dtypeobj_out))
+
+    return has_loss
+
+
+def dtype_limits(image, clip_negative=None):
```
Comment (sciunto) [inline skimage/util/dtype.py]: This is an API change, isn't it?

File: .gitignore [hunk 4]
```diff
@@ -32,3 +32,6 @@ doc/release/_release_notes_for_docs.txt
 doc/gh-pages
 wheels
 skimage/morphology/_skeletonize_3d_cy.pyx
+
+.ropeproject
```
Comment (sciunto) [inline .gitignore]: What's that?

File: skimage/util/dtype.py (hunk 3 above)
Comment (sciunto) [reply skimage/util/dtype.py]: No I read it changed from ` clip_negative=False)` to ` clip_negative=None)` The original code is `=False` as far as I can tell.

**Comment by hmaarrfk** [comment]
@sciunto i guess this one wont make 0.14 easily

https://github.com/scikit-image/scikit-image/blob/v0.14.x/skimage/util/dtype.py#L36

**Comment by sciunto** [comment]
yes, we changed the API because the former default value was dangerous (and if I remember well, caused a bug in the library).

**Comment by sciunto** [comment]
In test_downcast(), the warning differs from the one expected in the test.

**Comment by hmaarrfk** [comment]
@sciunto the `_scale` function is scary. I didn't realize it did [**magic**](https://github.com/scikit-image/scikit-image/blob/35ef2171e0da60e5d8f68c6c29b8e183a5c35a19/skimage/util/dtype.py#L176).

Although I really think this kind of refactoring pushes things in the right direction, as it stands, it isn't actually issuing the right error messages `img_as` is doing too much magic.

Lets wait until: https://github.com/scikit-image/scikit-image/issues/3373
this is resolved. Maybe after that one, I can make an audit of `img_as` for the int types.

**Comment by hmaarrfk** [comment]
I guess this is no longer necessary.
